File: pkg/limited/strategies.go
package limited

import (
	"strings"
	"time"
)

// FixedWindowStrategy implements fixed-window rate limiting.
type FixedWindowStrategy struct {
	WindowSize  time.Duration
	MaxRequests int

	IdentifierLimits map[string]int
	ResourceLimits   map[string]int
}

func NewFixedWindowStrategy(windowSize time.Duration, maxRequests int) *FixedWindowStrategy {
	return &FixedWindowStrategy{
		WindowSize:       windowSize,
		MaxRequests:      maxRequests,
		IdentifierLimits: make(map[string]int),
		ResourceLimits:   make(map[string]int),
	}
}

func (s *FixedWindowStrategy) CalculateWindows(now time.Time) []TimeWindow {
	if s.WindowSize <= 0 {
		return nil
	}

	windowStart := s.getWindowStart(now)
	windowEnd := windowStart.Add(s.WindowSize)

	return []TimeWindow{{
		Start: windowStart,
		End:   windowEnd,
		Key:   windowStart.Format(time.RFC3339),
	}}
}

func (s *FixedWindowStrategy) GetLimit(key RateLimitKey) int {
	if limit, ok := s.IdentifierLimits[key.Identifier]; ok {
		return limit
	}
	if limit, ok := s.ResourceLimits[key.Resource]; ok {
		return limit
	}
	return s.MaxRequests
}

func (s *FixedWindowStrategy) ShouldAllow(counts map[string]int, limit int) bool {
	total := 0
	for _, count := range counts {
		total += count
	}
	return total < limit
}

func (s *FixedWindowStrategy) getWindowStart(now time.Time) time.Time {
	windowNanos := s.WindowSize.Nanoseconds()
	if windowNanos <= 0 {
		return now
	}

	startNanos := (now.UnixNano() / windowNanos) * windowNanos
	return time.Unix(0, startNanos).In(now.Location())
}

func (s *FixedWindowStrategy) SetIdentifierLimit(identifier string, limit int) {
	s.IdentifierLimits[identifier] = limit
}

func (s *FixedWindowStrategy) SetResourceLimit(resource string, limit int) {
	s.ResourceLimits[resource] = limit
}

// SlidingWindowStrategy implements sliding-window rate limiting.
type SlidingWindowStrategy struct {
	WindowSize  time.Duration
	MaxRequests int
	Granularity time.Duration

	IdentifierLimits map[string]int
	ResourceLimits   map[string]int
}

func NewSlidingWindowStrategy(windowSize time.Duration, maxRequests int, granularity time.Duration) *SlidingWindowStrategy {
	return &SlidingWindowStrategy{
		WindowSize:       windowSize,
		MaxRequests:      maxRequests,
		Granularity:      granularity,
		IdentifierLimits: make(map[string]int),
		ResourceLimits:   make(map[string]int),
	}
}

func (s *SlidingWindowStrategy) CalculateWindows(now time.Time) []TimeWindow {
	if s.WindowSize <= 0 {
		return nil
	}

	granularity := s.Granularity
	if granularity <= 0 {
		granularity = time.Minute
	}

	subWindows := int(s.WindowSize / granularity)
	if subWindows < 1 {
		subWindows = 1
	}

	currentStart := now.Truncate(granularity)

	windows := make([]TimeWindow, 0, subWindows)
	for i := 0; i < subWindows; i++ {
		start := currentStart.Add(-time.Duration(i) * granularity)
		if now.Sub(start) > s.WindowSize {
			continue
		}
		end := start.Add(granularity)
		windows = append(windows, TimeWindow{
			Start: start,
			End:   end,
			Key:   start.Format(time.RFC3339),
		})
	}

	return windows
}

func (s *SlidingWindowStrategy) GetLimit(key RateLimitKey) int {
	if limit, ok := s.IdentifierLimits[key.Identifier]; ok {
		return limit
	}
	if limit, ok := s.ResourceLimits[key.Resource]; ok {
		return limit
	}
	return s.MaxRequests
}

func (s *SlidingWindowStrategy) ShouldAllow(counts map[string]int, limit int) bool {
	total := 0
	for _, count := range counts {
		total += count
	}
	return total < limit
}

func (s *SlidingWindowStrategy) SetIdentifierLimit(identifier string, limit int) {
	s.IdentifierLimits[identifier] = limit
}

func (s *SlidingWindowStrategy) SetResourceLimit(resource string, limit int) {
	s.ResourceLimits[resource] = limit
}

// MultiWindowStrategy enforces multiple limits simultaneously (for example: 100/min AND 1000/hour).
type MultiWindowStrategy struct {
	Windows          []WindowConfig
	IdentifierLimits map[string][]WindowConfig
	ResourceLimits   map[string][]WindowConfig
}

type WindowConfig struct {
	Duration    time.Duration
	MaxRequests int
}

func NewMultiWindowStrategy(windows []WindowConfig) *MultiWindowStrategy {
	return &MultiWindowStrategy{
		Windows:          append([]WindowConfig(nil), windows...),
		IdentifierLimits: make(map[string][]WindowConfig),
		ResourceLimits:   make(map[string][]WindowConfig),
	}
}

func (s *MultiWindowStrategy) CalculateWindows(now time.Time) []TimeWindow {
	if len(s.Windows) == 0 {
		return nil
	}

	windows := make([]TimeWindow, 0, len(s.Windows))
	for _, config := range s.Windows {
		if config.Duration <= 0 {
			continue
		}
		window := GetFixedWindow(now, config.Duration)
		windows = append(windows, TimeWindow{
			Start: window.Start,
			End:   window.End,
			Key:   window.Start.Format(time.RFC3339) + "_" + config.Duration.String(),
		})
	}
	return windows
}

func (s *MultiWindowStrategy) GetLimit(key RateLimitKey) int {
	limits := s.limitsForKey(key)
	if len(limits) == 0 {
		return 0
	}
	return limits[0].MaxRequests
}

func (s *MultiWindowStrategy) ShouldAllow(counts map[string]int, limit int) bool {
	if len(s.Windows) == 0 {
		return false
	}

	// Map counts deterministically by duration suffix (mirrors CalculateWindows keys).
	for _, config := range s.Windows {
		if config.Duration <= 0 {
			continue
		}
		max := config.MaxRequests
		suffix := "_" + config.Duration.String()

		count := 0
		for key, observed := range counts {
			if strings.HasSuffix(key, suffix) {
				count = observed
				break
			}
		}

		if count >= max {
			return false
		}
	}

	return true
}

func (s *MultiWindowStrategy) limitsForKey(key RateLimitKey) []WindowConfig {
	if limits, ok := s.IdentifierLimits[key.Identifier]; ok && len(limits) > 0 {
		return limits
	}
	if limits, ok := s.ResourceLimits[key.Resource]; ok && len(limits) > 0 {
		return limits
	}
	return s.Windows
}

File: ts/src/limited/strategies.ts
import { getFixedWindow, formatWindowId } from "./models.js";
import type { RateLimitKey, RateLimitStrategy, TimeWindow } from "./types.js";

export class FixedWindowStrategy implements RateLimitStrategy {
  readonly windowSizeMs: number;
  readonly maxRequests: number;
  readonly identifierLimits: Record<string, number>;
  readonly resourceLimits: Record<string, number>;

  constructor(windowSizeMs: number, maxRequests: number) {
    this.windowSizeMs = Math.floor(Number(windowSizeMs));
    this.maxRequests = Math.floor(Number(maxRequests));
    this.identifierLimits = {};
    this.resourceLimits = {};
  }

  calculateWindows(now: Date): TimeWindow[] {
    const size = Math.floor(this.windowSizeMs);
    if (!Number.isFinite(size) || size <= 0) {
      return [];
    }

    const startMs = Math.floor(now.valueOf() / size) * size;
    const start = new Date(startMs);
    const end = new Date(startMs + size);

    return [{ start, end, key: formatWindowId(start) }];
  }

  getLimit(key: RateLimitKey): number {
    const id = String(key.identifier ?? "");
    if (id && id in this.identifierLimits)
      return this.identifierLimits[id] ?? 0;
    const res = String(key.resource ?? "");
    if (res && res in this.resourceLimits) return this.resourceLimits[res] ?? 0;
    return this.maxRequests;
  }

  shouldAllow(counts: Record<string, number>, limit: number): boolean {
    let total = 0;
    for (const v of Object.values(counts)) {
      total += Number(v) || 0;
    }
    return total < Number(limit);
  }

  setIdentifierLimit(identifier: string, limit: number): void {
    this.identifierLimits[String(identifier)] = Math.floor(Number(limit));
  }

  setResourceLimit(resource: string, limit: number): void {
    this.resourceLimits[String(resource)] = Math.floor(Number(limit));
  }
}

export class SlidingWindowStrategy implements RateLimitStrategy {
  readonly windowSizeMs: number;
  readonly maxRequests: number;
  readonly granularityMs: number;
  readonly identifierLimits: Record<string, number>;
  readonly resourceLimits: Record<string, number>;

  constructor(
    windowSizeMs: number,
    maxRequests: number,
    granularityMs: number,
  ) {
    this.windowSizeMs = Math.floor(Number(windowSizeMs));
    this.maxRequests = Math.floor(Number(maxRequests));
    this.granularityMs = Math.floor(Number(granularityMs));
    this.identifierLimits = {};
    this.resourceLimits = {};
  }

  calculateWindows(now: Date): TimeWindow[] {
    const windowMs = Math.floor(this.windowSizeMs);
    if (!Number.isFinite(windowMs) || windowMs <= 0) return [];

    let granularity = Math.floor(this.granularityMs);
    if (!Number.isFinite(granularity) || granularity <= 0) granularity = 60_000;

    let subWindows = Math.floor(windowMs / granularity);
    if (subWindows < 1) subWindows = 1;

    const nowMs = now.valueOf();
    const currentStartMs = Math.floor(nowMs / granularity) * granularity;

    const windows: TimeWindow[] = [];
    for (let i = 0; i < subWindows; i += 1) {
      const startMs = currentStartMs - i * granularity;
      if (nowMs - startMs > windowMs) continue;
      const start = new Date(startMs);
      const end = new Date(startMs + granularity);
      windows.push({ start, end, key: formatWindowId(start) });
    }
    return windows;
  }

  getLimit(key: RateLimitKey): number {
    const id = String(key.identifier ?? "");
    if (id && id in this.identifierLimits)
      return this.identifierLimits[id] ?? 0;
    const res = String(key.resource ?? "");
    if (res && res in this.resourceLimits) return this.resourceLimits[res] ?? 0;
    return this.maxRequests;
  }

  shouldAllow(counts: Record<string, number>, limit: number): boolean {
    let total = 0;
    for (const v of Object.values(counts)) {
      total += Number(v) || 0;
    }
    return total < Number(limit);
  }

  setIdentifierLimit(identifier: string, limit: number): void {
    this.identifierLimits[String(identifier)] = Math.floor(Number(limit));
  }

  setResourceLimit(resource: string, limit: number): void {
    this.resourceLimits[String(resource)] = Math.floor(Number(limit));
  }
}

export type WindowConfig = { durationMs: number; maxRequests: number };

export class MultiWindowStrategy implements RateLimitStrategy {
  readonly windows: WindowConfig[];
  readonly identifierLimits: Record<string, WindowConfig[]>;
  readonly resourceLimits: Record<string, WindowConfig[]>;

  constructor(windows: WindowConfig[]) {
    this.windows = Array.isArray(windows) ? windows.map((w) => ({ ...w })) : [];
    this.identifierLimits = {};
    this.resourceLimits = {};
  }

  calculateWindows(now: Date): TimeWindow[] {
    if (this.windows.length === 0) return [];

    const out: TimeWindow[] = [];
    for (const cfg of this.windows) {
      const durationMs = Math.floor(Number(cfg.durationMs));
      if (!Number.isFinite(durationMs) || durationMs <= 0) continue;
      const window = getFixedWindow(now, durationMs);
      out.push({
        start: window.start,
        end: window.end,
        key: `${formatWindowId(window.start)}_${durationMs}ms`,
      });
    }
    return out;
  }

  getLimit(key: RateLimitKey): number {
    const limits = this._limitsForKey(key);
    if (limits.length === 0) return 0;
    return Math.floor(Number(limits[0]?.maxRequests ?? 0));
  }

  shouldAllow(counts: Record<string, number>, _limit: number): boolean {
    if (this.windows.length === 0) return false;

    for (const cfg of this.windows) {
      const durationMs = Math.floor(Number(cfg.durationMs));
      if (!Number.isFinite(durationMs) || durationMs <= 0) continue;
      const suffix = `_${durationMs}ms`;

      let count = 0;
      for (const [key, observed] of Object.entries(counts)) {
        if (String(key).endsWith(suffix)) {
          count = Math.floor(Number(observed) || 0);
          break;
        }
      }

      const maxAllowed = Math.floor(Number(cfg.maxRequests) || 0);
      if (count >= maxAllowed) return false;
    }

    return true;
  }

  private _limitsForKey(key: RateLimitKey): WindowConfig[] {
    const id = String(key.identifier ?? "");
    if (id && id in this.identifierLimits) {
      const v = this.identifierLimits[id];
      if (Array.isArray(v) && v.length > 0) return v;
    }
    const res = String(key.resource ?? "");
    if (res && res in this.resourceLimits) {
      const v = this.resourceLimits[res];
      if (Array.isArray(v) && v.length > 0) return v;
    }
    return this.windows;
  }
}
